// shoukaku/dist/src/node/Queue.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queue = void 0;
const ws_1 = __importDefault(require("ws"));
/**
 * Represents a message queue
 * @internal
 */
class Queue {
    /**
     * @param node An instance of Node
     */
    constructor(node) {
        this.node = node;
        this.pending = [];
        this.flushes = 0;
    }
    /**
     * Add data to queue
     * @param data Message data
     * @param important Priority status
     * @internal
     */
    add(data, important = false) {
        if (data)
            this.pending[important ? 'unshift' : 'push'](JSON.stringify(data));
        this.process();
    }
    /**
     * Clear the queue
     * @internal
     */
    clear() {
        this.pending.length = 0;
    }
    /**
     * Flush the ws connections in queue
     * @param code Status code
     * @param reason Reason for close
     * @internal
     */
    flush(code, reason) {
        if (!this.pending.length || this.flushes > 10) {
            this.flushes = 0;
            this.clear();
            this.node.ws?.close(code, reason);
            return;
        }
        this.flushes++;
        setTimeout(() => this.flush(code, reason), 1000);
    }
    /**
     * Process messages in the queue
     * @internal
     */
    process() {
        if (!this.node.ws || this.node.ws.readyState !== ws_1.default.OPEN || !this.pending.length)
            return;
        while (this.pending.length) {
            const message = this.pending.shift();
            if (!message)
                return;
            this.node.ws.send(message);
        }
    }
}
exports.Queue = Queue;
