// shoukaku/dist/src/guild/Player.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Player = exports.Filters = void 0;
const events_1 = require("events");
const Utils_1 = require("../Utils");
const Constants_1 = require("../Constants");
const Connection_1 = require("./Connection");
/**
 * Lavalink filters
 */
class Filters {
    /**
     * Options to initialize this filters instance with
     * @param options.volume The volume to play audio at as a decimal
     * @param options.equalizer An array of objects that conforms to the Bands type that define volumes at different frequencies
     * @param options.karaoke An object that conforms to the KaraokeSettings type that defines a range of frequencies to mute
     * @param options.timescale An object that conforms to the TimescaleSettings type that defines the time signature to play the audio at
     * @param options.tremolo An object that conforms to the FreqSettings type that defines an ocillation in volume
     * @param options.vibrato An object that conforms to the FreqSettings type that defines an ocillation in pitch
     * @param options.rotation An object that conforms to the RotationSettings type that defines the frequency of audio rotating round the listener
     * @param options.distortion An object that conforms to DistortionSettings that defines distortions in the audio
     * @param options.channelMix An object that conforms to ChannelMixSettings that defines how much the left and right channels affect each other (setting all factors to 0.5 causes both channels to get the same audio)
     * @param options.lowPass An object that conforms to LowPassSettings that defines the amount of suppression on higher frequencies
     */
    constructor(options = {}) {
        this.volume = options.volume ?? 1.0;
        this.equalizer = options.equalizer || [];
        this.karaoke = options.karaoke || null;
        this.timescale = options.timescale || null;
        this.tremolo = options.tremolo || null;
        this.vibrato = options.vibrato || null;
        this.rotation = options.rotation || null;
        this.distortion = options.distortion || null;
        this.channelMix = options.channelMix || null;
        this.lowPass = options.lowPass || null;
    }
}
exports.Filters = Filters;
/**
 * Wrapper object around Lavalink
 */
class Player extends events_1.EventEmitter {
    /**
     * @param node An instance of Node (Lavalink API wrapper)
     * @param options.guildId Guild ID in which voice channel to connect to is located
     * @param options.shardId Shard ID in which the guild exists
     * @param options.channelId Channel ID of voice channel to connect to
     * @param options.deaf Optional boolean value to specify whether to deafen the current bot user
     * @param options.mute Optional boolean value to specify whether to mute the current bot user
     */
    constructor(node, options) {
        super();
        this.node = node;
        this.connection = new Connection_1.Connection(this, options);
        this.track = null;
        this.paused = false;
        this.position = 0;
        this.ping = 0;
        this.filters = new Filters();
    }
    /**
     * Move player to another node
     * @param name Name of node to move to
     * @returns The current player instance
     */
    move(name) {
        const node = this.node.manager.nodes.get(name);
        if (!node || node.name === this.node.name)
            return this;
        if (node.state !== Constants_1.State.CONNECTED)
            throw new Error('The node you specified is not ready');
        this.connection.destroyLavalinkPlayer();
        this.node.players.delete(this.connection.guildId);
        this.node = node;
        this.node.players.set(this.connection.guildId, this);
        this.connection.resendServerUpdate();
        this.resume();
        return this;
    }
    /**
     * Play a new track
     * @param playable Options for playing this track
     * @returns The current player instance
     */
    playTrack(playable) {
        const payload = {
            op: Constants_1.OPCodes.PLAY,
            guildId: this.connection.guildId,
            track: playable.track,
            noReplace: playable.options?.noReplace ?? true,
            pause: playable.options?.pause ?? false
        };
        if (playable.options?.startTime)
            payload.startTime = playable.options.startTime;
        if (playable.options?.endTime)
            payload.endTime = playable.options.startTime;
        this.node.queue.add(payload);
        this.track = playable.track;
        this.paused = playable.options?.pause ?? false;
        this.position = 0;
        return this;
    }
    /**
     * Stop the currently playing track
     * @returns The current player instance
     */
    stopTrack() {
        this.position = 0;
        this.node.queue.add({
            op: Constants_1.OPCodes.STOP,
            guildId: this.connection.guildId
        });
        return this;
    }
    /**
     * Pause or unpause the currently playing track
     * @param pause Boolean value to specify whether to pause or unpause the current bot user
     * @returns The current player instance
     */
    setPaused(pause = true) {
        this.node.queue.add({
            op: Constants_1.OPCodes.PAUSE,
            guildId: this.connection.guildId,
            pause
        });
        this.paused = pause;
        return this;
    }
    /**
     * Seek to a specific time in the currently playing track
     * @param position Position to seek to in milliseconds
     * @returns The current player instance
     */
    seekTo(position) {
        this.node.queue.add({
            op: Constants_1.OPCodes.SEEK,
            guildId: this.connection.guildId,
            position
        });
        return this;
    }
    /**
     * Change the volume of the currently playing track
     * @param volume Target volume as a decimal
     * @returns The current player instance
     */
    setVolume(volume) {
        volume = Math.min(5, Math.max(0, volume));
        this.filters.volume = volume;
        this.updateFilters();
        return this;
    }
    /**
     * Change the equalizer settings applied to the currently playing track
     * @param bands An array of objects that conforms to the Bands type that define volumes at different frequencies
     * @returns The current player instance
     */
    setEqualizer(bands) {
        this.filters.equalizer = bands;
        this.updateFilters();
        return this;
    }
    /**
     * Change the karaoke settings applied to the currently playing track
     * @param karaoke An object that conforms to the KaraokeSettings type that defines a range of frequencies to mute
     * @returns The current player instance
     */
    setKaraoke(karaoke) {
        this.filters.karaoke = karaoke || null;
        this.updateFilters();
        return this;
    }
    /**
     * Change the timescale settings applied to the currently playing track
     * @param timescale An object that conforms to the TimescaleSettings type that defines the time signature to play the audio at
     * @returns The current player instance
     */
    setTimescale(timescale) {
        this.filters.timescale = timescale || null;
        this.updateFilters();
        return this;
    }
    /**
     * Change the tremolo settings applied to the currently playing track
     * @param tremolo An object that conforms to the FreqSettings type that defines an ocillation in volume
     * @returns The current player instance
     */
    setTremolo(tremolo) {
        this.filters.tremolo = tremolo || null;
        this.updateFilters();
        return this;
    }
    /**
     * Change the vibrato settings applied to the currently playing track
     * @param vibrato An object that conforms to the FreqSettings type that defines an ocillation in pitch
     * @returns The current player instance
     */
    setVibrato(vibrato) {
        this.filters.vibrato = vibrato || null;
        this.updateFilters();
        return this;
    }
    /**
     * Change the rotation settings applied to the currently playing track
     * @param rotation An object that conforms to the RotationSettings type that defines the frequency of audio rotating round the listener
     * @returns The current player instance
     */
    setRotation(rotation) {
        this.filters.rotation = rotation || null;
        this.updateFilters();
        return this;
    }
    /**
     * Change the distortion settings applied to the currently playing track
     * @param distortion An object that conforms to DistortionSettings that defines distortions in the audio
     * @returns The current player instance
     */
    setDistortion(distortion) {
        this.filters.distortion = distortion || null;
        this.updateFilters();
        return this;
    }
    /**
     * Change the channel mix settings applied to the currently playing track
     * @param mix An object that conforms to ChannelMixSettings that defines how much the left and right channels affect each other (setting all factors to 0.5 causes both channels to get the same audio)
     * @returns The current player instance
     */
    setChannelMix(mix) {
        this.filters.channelMix = mix || null;
        this.updateFilters();
        return this;
    }
    /**
     * Change the low pass settings applied to the currently playing track
     * @param pass An object that conforms to LowPassSettings that defines the amount of suppression on higher frequencies
     * @returns The current player instance
     */
    setLowPass(pass) {
        this.filters.lowPass = pass || null;
        this.updateFilters();
        return this;
    }
    /**
     * Change the all filter settings applied to the currently playing track
     * @param options An object that conforms to FilterOptions that defines all filters to apply/modify
     * @returns The current player instance
     */
    setFilters(options) {
        this.filters = new Filters(options);
        this.updateFilters();
        return this;
    }
    /**
     * Clear all filters applied to the currently playing track
     * @returns The current player instance
     */
    clearFilters() {
        this.filters = new Filters();
        this.node.queue.add({
            op: Constants_1.OPCodes.FILTERS,
            guildId: this.connection.guildId
        });
        return this;
    }
    /**
     * Resume the current track
     * @param options An object that conforms to ResumeOptions that specify behavior on resuming
     * @returns The current player instance
     */
    resume(options = {}) {
        this.updateFilters();
        if (this.track) {
            options = (0, Utils_1.mergeDefault)({ startTime: this.position, pause: this.paused }, options);
            this.playTrack({ track: this.track, options });
        }
        this.emit('resumed');
        return this;
    }
    /**
     * Update all filters via the filters operation
     * @internal
     */
    updateFilters() {
        const { volume, equalizer, karaoke, timescale, tremolo, vibrato, rotation, distortion, channelMix, lowPass } = this.filters;
        this.node.queue.add({
            op: Constants_1.OPCodes.FILTERS,
            guildId: this.connection.guildId,
            volume,
            equalizer,
            karaoke,
            timescale,
            tremolo,
            vibrato,
            rotation,
            distortion,
            channelMix,
            lowPass
        });
    }
    /**
     * Remove all event listeners on this instance
     * @internal
     */
    clean() {
        this.removeAllListeners();
        this.reset();
    }
    /**
     * Reset the track, position and filters on this instance to defaults
     */
    reset() {
        this.track = null;
        this.position = 0;
        this.filters = new Filters();
    }
    /**
     * Handle JSON data recieved from Lavalink
     * @param json JSON data from Lavalink
     * @internal
     */
    onLavalinkMessage(json) {
        if (json.op === Constants_1.OPCodes.PLAYER_UPDATE) {
            this.position = json.state.position;
            // ping property require lavalink >=3.5.1
            this.ping = json.state.ping ?? 0;
            this.emit('update', json);
        }
        else if (json.op === Constants_1.OPCodes.EVENT)
            this.onPlayerEvent(json);
        else {
            this.node.emit('debug', this.node.name, `[Player] -> [Node] : Unknown Message OP ${json.op} | Guild: ${this.connection.guildId}`);
        }
    }
    /**
     * Handle player events recieved from Lavalink
     * @param json JSON data from Lavalink
     * @internal
     */
    onPlayerEvent(json) {
        switch (json.type) {
            case 'TrackStartEvent':
                this.position = 0;
                this.emit('start', json);
                break;
            case 'TrackEndEvent':
                this.emit('end', json);
                break;
            case 'TrackStuckEvent':
                this.emit('stuck', json);
                break;
            case 'TrackExceptionEvent':
                this.emit('exception', json);
                break;
            case 'WebSocketClosedEvent':
                if (!this.connection.reconnecting) {
                    if (!this.connection.moved)
                        this.emit('closed', json);
                    else
                        this.connection.moved = false;
                }
                break;
            default:
                this.node.emit('debug', this.node.name, `[Player] -> [Node] : Unknown Player Event Type ${json.type} | Guild: ${this.connection.guildId}`);
        }
    }
}
exports.Player = Player;
