// shoukaku/dist/src/Constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeDefaults = exports.ShoukakuDefaults = exports.OPCodes = exports.VoiceState = exports.State = void 0;
const package_json_1 = __importDefault(require("../package.json"));
var State;
(function (State) {
    State[State["CONNECTING"] = 0] = "CONNECTING";
    State[State["CONNECTED"] = 1] = "CONNECTED";
    State[State["DISCONNECTING"] = 2] = "DISCONNECTING";
    State[State["DISCONNECTED"] = 3] = "DISCONNECTED";
})(State = exports.State || (exports.State = {}));
var VoiceState;
(function (VoiceState) {
    VoiceState[VoiceState["SESSION_READY"] = 0] = "SESSION_READY";
    VoiceState[VoiceState["SESSION_ID_MISSING"] = 1] = "SESSION_ID_MISSING";
    VoiceState[VoiceState["SESSION_ENDPOINT_MISSING"] = 2] = "SESSION_ENDPOINT_MISSING";
})(VoiceState = exports.VoiceState || (exports.VoiceState = {}));
var OPCodes;
(function (OPCodes) {
    // From Lavalink
    OPCodes["VOICE_UPDATE"] = "voiceUpdate";
    OPCodes["PLAY"] = "play";
    OPCodes["STOP"] = "stop";
    OPCodes["PAUSE"] = "pause";
    OPCodes["SEEK"] = "seek";
    OPCodes["VOLUME"] = "volume";
    OPCodes["FILTERS"] = "filters";
    OPCodes["DESTROY"] = "destroy";
    // To Lavalink
    OPCodes["PLAYER_UPDATE"] = "playerUpdate";
    OPCodes["CONFIGURE_RESUMING"] = "configureResuming";
    OPCodes["STATS"] = "stats";
    OPCodes["EVENT"] = "event";
})(OPCodes = exports.OPCodes || (exports.OPCodes = {}));
exports.ShoukakuDefaults = {
    resume: false,
    resumeKey: `Shoukaku@${package_json_1.default.version}(${package_json_1.default.repository.url})`,
    resumeTimeout: 30000,
    resumeByLibrary: false,
    alwaysSendResumeKey: false,
    reconnectTries: 3,
    reconnectInterval: 5000,
    restTimeout: 60000,
    moveOnDisconnect: false,
    userAgent: `${package_json_1.default.name}bot/${package_json_1.default.version} (${package_json_1.default.repository.url})`,
    structures: {}
};
exports.NodeDefaults = {
    name: 'Default',
    url: '',
    auth: '',
    secure: false,
    group: undefined
};
